File: std/list/sorting.hs
import std/list/bound-checks
import std/list/functional

-- Sorting
partition sort(list):
        less <- filter(list, x -> x <  list[0]),
        more <- filter(list, x -> x >= list[0])
->
        empty(list) ? []
                    : partition sort(less) ~ list[0] ~ partition sort(more)

File: std/array/sorting.f
import std/array/functional

== Sorts given `array` in the ascending order.
==
== Returns: Array
==     Array of sorted elements from `array`.
sort (array) {
    end << array = [],
    less << end ? [] : filter (array, x -> x <  array[0]),
    more << end ? [] : filter (array, x -> x >= array[0])
} ->
    end ? [] : sort (less) ~ array[0] ~ sort (more)